### ISSUE TITLE
r10k deletes the contents of /etc/puppet/environments if the sources:local:remote address is invalid

### ISSUE THREAD

**Issue by Resisty** [puppetlabs/r10k#193]
I was testing out r10k, having heard good things, and attempted to deploy a new test environment, 'bobi':

```
# /usr/local/bin/r10k deploy environment bobi -p
Faraday: you may want to install system_timer for reliable timeouts
[R10K::TaskRunner - ERROR] Task #<R10K::Task::Deployment::DeployEnvironments:0x7f9306332250> failed while running: Couldn't update git cache for https://github.com/Resisty/r10k_environment: "fatal: could not read Username for 'https://github.com': No such device or address"
# vim /etc/r10k.yaml 
# /usr/local/bin/r10k deploy environment bobi -p
Faraday: you may want to install system_timer for reliable timeouts

Error while running: #<Errno::ENOENT: No such file or directory - getcwd>
# pwd
pwd: couldn't find directory entry in `..' with matching i-node
# cd ..
# pwd
pwd: couldn't find directory entry in `..' with matching i-node
# cd ..
# pwd
pwd: couldn't find directory entry in `..' with matching i-node
# cd ..
# pwd 
pwd: couldn't find directory entry in `..' with matching i-node
# cd ..
# pwd
/etc/puppet/environments
# ls -al
total 8
drwxr-xr-x  2 puppet puppet 4096 Oct  8 09:17 .
drwxr-xr-x 12 puppet puppet 4096 Oct  6 12:37 ..

```

As you can see, everything under /etc/puppet/environments was removed. Previously there was a production and several test environments.

Fortunately, I have a backup of /etc/puppet/environments/production, but this is a pretty egregious error.

**Comment by adrienthebo** [comment]
Which version of r10k are you using? https://github.com/adrienthebo/r10k/issues/178 sounds similar and was fixed in 1.3.4.

**Comment by Resisty** [comment]
I'm using 1.3.4, just installed it from `gem install r10k` this morning:

```
# /usr/local/bin/r10k version
Faraday: you may want to install system_timer for reliable timeouts
1.3.4
```

When I get a chance, I'll post the `r10k deploy environment -v debug2 --trace` here; still making sure the environments restored OK.

**Comment by bunkzilla** [comment]
I noticed the same behavior, I downloaded 1.3.4, but same behavior is being exhibited. 

```
 r10k deploy environment -v  debug2 --trace
Faraday: you may want to install system_timer for reliable timeouts
[R10K::Source::Git - INFO] Determining current branches for "https://glithub.com/glarizza/puppet_repository.git"
[R10K::Git::Cache - DEBUG] Creating new git cache for "https://glithub.com/glarizza/puppet_repository.git"
[R10K::Git::Cache - DEBUG1] Execute: git clone --mirror https://glithub.com/glarizza/puppet_repository.git /var/cache/r10k/https---glithub.com-glarizza-puppet_repository.git
[R10K::Git::Cache - DEBUG2] [git clone --mirror https://glithub.com/glarizza/puppet_repository.git /var/cache/r10k/https---glithub.com-glarizza-puppet_repository.git] STDOUT: Initialized empty Git repository in /var/cache/r10k/https---glithub.com-glarizza-puppet_repository.git/
[R10K::Git::Cache - DEBUG2] [git clone --mirror https://glithub.com/glarizza/puppet_repository.git /var/cache/r10k/https---glithub.com-glarizza-puppet_repository.git] STDERR: error: SSL: certificate subject name 'ssl2000.cloudflare.com' does not match target host name 'glithub.com' while accessing https://glithub.com/glarizza/puppet_repository.git/info/refs

fatal: HTTP request failed
[R10K::TaskRunner - ERROR] Task #<R10K::Task::Deployment::DeployEnvironments:0x7fb5c5260c78> failed while running: Couldn't update git cache for https://glithub.com/glarizza/puppet_repository.git: "fatal: HTTP request failed"
/usr/lib/ruby/gems/1.8/gems/r10k-1.3.4/lib/r10k/git/cache.rb:73:in `sync!'
/usr/lib/ruby/gems/1.8/gems/r10k-1.3.4/lib/r10k/git/cache.rb:52:in `sync'
/usr/lib/ruby/gems/1.8/gems/r10k-1.3.4/lib/r10k/source/git.rb:67:in `preload!'
/usr/lib/ruby/gems/1.8/gems/log4r-1.1.10/lib/log4r/outputter/outputterfactory.rb:50:in `to_proc'
/usr/lib/ruby/gems/1.8/gems/r10k-1.3.4/lib/r10k/deployment.rb:28:in `each'
/usr/lib/ruby/gems/1.8/gems/r10k-1.3.4/lib/r10k/deployment.rb:28:in `preload!'
/usr/lib/ruby/gems/1.8/gems/r10k-1.3.4/lib/r10k/task/deployment.rb:76:in `call'
/usr/lib/ruby/gems/1.8/gems/r10k-1.3.4/lib/r10k/task_runner.rb:26:in `run'
/usr/lib/ruby/gems/1.8/gems/r10k-1.3.4/lib/r10k/task_runner.rb:21:in `catch'
/usr/lib/ruby/gems/1.8/gems/r10k-1.3.4/lib/r10k/task_runner.rb:21:in `run'
/usr/lib/ruby/gems/1.8/gems/r10k-1.3.4/lib/r10k/cli/deploy.rb:67:in `command'
/usr/lib/ruby/gems/1.8/gems/cri-2.5.0/lib/cri/command.rb:298:in `call'
/usr/lib/ruby/gems/1.8/gems/cri-2.5.0/lib/cri/command.rb:298:in `run_this'
/usr/lib/ruby/gems/1.8/gems/cri-2.5.0/lib/cri/command.rb:251:in `run'
/usr/lib/ruby/gems/1.8/gems/cri-2.5.0/lib/cri/command.rb:264:in `run'
/usr/lib/ruby/gems/1.8/gems/r10k-1.3.4/bin/r10k:7
/usr/bin/r10k:19:in `load'
/usr/bin/r10k:19
[R10K::Task::Deployment::PurgeEnvironments - INFO] Purging stale environments from /etc/puppet/environments

r10k version
Faraday: you may want to install system_timer for reliable timeouts
1.3.4

```

**Comment by adrienthebo** [comment]
@bunkzilla the logging indicates that the initial repository clone failed; in addition to that did you see existing environments removed as well?

**Comment by bunkzilla** [comment]
Yes, what I did was first run the command with github properly spelled in the /etc/r10k.yaml file.   It added webinar_env and production.

I then introduced the typo of glithub.   I reran the deploy to see what would happen.  

Yes, it removed the environments.   I was thankful i was running in a sandbox.  :)

**Comment by adrienthebo** [comment]
@bunkzilla if you change the remote that doesn't exist there's not a lot that can be done; if the remote is not reachable then the cached repository can be used to list branches and r10k can continue. However if you rename the remote then the cached remote can't be used at all, r10k can't use existing information to determine what branches should exist, and will do what you've seen. I would consider that as working by design; it may not be desirable behavior but it's not actively buggy.

However if r10k fails to fetch an existing remote due to transient issues and  removes all branches regardless, that's a different issue.

**Comment by bunkzilla** [comment]
That's great news.   Sorry to pollute the thread with my invalid test case.

**Comment by adrienthebo** [comment]
@bunkzilla no worries; I think that you bring up a good case - if a source can't be fetched then it's no longer safe to remove unmanaged directories. I think that you should file that as a new issue that we can address separately.

**Comment by adrienthebo** [comment]
GH-199 should provide a fix for this by coupling environment preloading and initial generation and environment purging. If the former fails then further execution will be halted, preventing any purging of environments.

**Comment by adrienthebo** [comment]
@bunkzilla also, you were correct the first time - your issue is the same as this. I misread the initial issue as inadvertently an already-fetched environment but the issue was that an environment that had never been fetched, as you demonstrated. I'm going to close out GH-194 in favor of this issue, but thank you for helping reproduce this issue and adding more information!

**Comment by adrienthebo** [comment]
This should be fixed as part of GH-199, it would be a great help if someone could verify that this solves the issue as expected. If I don't hear anymore issues in about a week I'll consider this issue as fixed in master and will release it in 1.4.0.

**Comment by adrienthebo** [comment]
Since there were no additional comments I'm considering this fixed; this will be released in 1.4.0. @Resisty @bunkzilla thanks for reporting this!
